Lab-3:  Синцов Михаил Юрьевич 6201-120303D

dadppil opened the pull request.

dadppil committed:
Add files via upload

DiBoS290699 reviewed (commented):
# Результаты проверки текста отчета
## Рецензия на отчёт по лабораторной работе №3

**1. Краткое резюме**
Отчёт демонстрирует хорошее понимание принципов обработки исключений и реализации связных списков.  В целом, работа выполнена успешно, все поставленные задачи охвачены. Ключевой плюс – корректная реализация исключений и их обработка. Главный риск/пробел – отсутствие тестов, подтверждающих корректную работу всех сценариев, особенно в `LinkedListTabulatedFunction`.

**2. Соответствие заданию**
*   **Да**. Отчёт полностью соответствует заданию.
    *   Все классы исключений созданы и задокументированы.
    *   Методы `TabulatedFunction` реализованы с выбрасыванием соответствующих исключений в указанных условиях.
    *   Реализация на основе связного списка выполнена.
    *   Создан интерфейс `TabulatedFunction`.
*   "Оба конструктора класса должны выбрасывать исключение `IllegalArgumentException`, если левая граница области определения больше или равна правой, а также если предлагаемое количество точек меньше двух." – Реализовано.
*   "Методы `getPoint()`, `setPoint()`, `getPointX()`, `setPointX()`, `getPointY()`, `setPointY()` и `deletePoint()` должны выбрасывать исключение `FunctionPointIndexOutOfBoundsException`, если переданный в метод номер выходит за границы набора точек." – Реализовано.
*   "Методы `setPoint()` и `setPointX()` должны выбрасывать исключение `InappropriateFunctionPointException` в том случае, если координата `x` задаваемой точки лежит вне интервала, определяемого значениями соседних точек табулированной функции." – Реализовано.

**3. Техническая корректность (ООП и Java)**
*   **Инкапсуляция:** В `LinkedListTabulatedFunction` реализована инкапсуляция внутреннего списка, публичные методы не возвращают ссылки на узлы списка.
*   **Абстракция:** Интерфейс `TabulatedFunction` абстрагирует общие операции над табулированными функциями.
*   **Наследование:** Классы исключений наследуются от соответствующих базовых классов (`IndexOutOfBoundsException`, `Exception`).
*   **Иерархии/интерфейсы:** Иерархия исключений простая, но достаточная для поставленной задачи. Интерфейс `TabulatedFunction` позволяет работать с разными реализациями полиморфно.
*   **Обработка исключений:** Все исключения выбрасываются в соответствии с заданием, сообщения об ошибках информативны.
*   **По описанию**:  Механизм сравнения вещественных чисел с использованием `EPSILON` корректен и оправдан.

**4. Архитектура и дизайн**
*   **Модульная структура:** Пакет `functions` логически разделяет классы, связанные с табулированными функциями.
*   **Паттерны:** Композиция между `LinkedListTabulatedFunction` и `FunctionNode` реализована корректно.
*   **Обоснование решений:**  Объяснено использование связного списка, инкапсуляции и машинного эпсилона.
*   **UML-диаграммы:** Отсутствуют — допустимо, но могли бы помочь в понимании структуры `LinkedListTabulatedFunction`.

**5. Воспроизводимость (опционально)**
*   Не представлено — допустимо.

**6. Тестирование и результаты (опционально)**
*   Не представлено — допустимо.  Отсутствие тестов является основным недостатком.  Необходимо добавить модульные тесты для проверки всех методов, особенно в `LinkedListTabulatedFunction`.

**7. Стиль и конвенции (применимо при наличии кода)**
*   **Нейминг:** Имена классов, методов и переменных соответствуют Java-конвенциям.
*   **Javadoc:** Отсутствует — допустимо, но рекомендуется добавить для публичных API.
*   **Форматирование:** Код форматирован читабельно.

**8. Источники и добросовестность**
*   Ссылки на внешние материалы отсутствуют, что допустимо для данной задачи.  Подозрение на заимствования отсутствует.

**9. Рекомендации по улучшению**
*   Добавьте Javadoc для всех публичных методов, чтобы улучшить читабельность и понимание API.
*   Напишите модульные тесты для `ArrayTabulatedFunction` и `LinkedListTabulatedFunction`, покрывающие все сценарии, включая граничные случаи и обработку исключений.
*   Рассмотрите возможность добавления UML-диаграммы класса `LinkedListTabulatedFunction` для визуализации структуры и отношений между классами.
*   Обоснуйте выбор двусвязного циклически списка. Какие альтернативные структуры данных рассматривались и почему был выбран именно этот вариант?
*   Если будете добавлять тесты, используйте JUnit или аналогичный фреймворк для автоматизации процесса тестирования.
*   Улучшите ясность изложения, добавив больше комментариев в коде, объясняющих сложные алгоритмы и логику работы.


# Результаты проверки текста отчета на плагиаризм
Обнаружены значимые совпадения в фрагментах кода, связанных с работой с функциями и обработкой исключений. Наибольшее сходство наблюдается с работами seregaqt и sharllottka, а также sr1racha777. Вероятно, имеет место заимствование отдельных фрагментов кода, реализующих схожую логику.

# Результаты проверки кода
## Рецензия на лабораторную работу

**Работоспособность:**

В целом, код соответствует заданию. Реализованы классы `ArrayTabulatedFunction` и `LinkedListTabulatedFunction`, а также необходимые исключения. Функциональность обеих реализаций табулированных функций соответствует требованиям: добавление, удаление, изменение точек, получение значений, обработка исключительных ситуаций.  Однако, в `LinkedListTabulatedFunction` есть потенциальные проблемы с производительностью при частом обращении к элементам списка, так как не используется эффективное кэширование.  В `ArrayTabulatedFunction` при добавлении элемента реализовано расширение массива, что правильно, но могло бы быть более оптимальным по памяти.

**Пример проблемного участка кода (LinkedListTabulatedFunction):**

```java
 private FunctionNode getNodeByIndex(int index) {
        if (index < 0 || index >= pointsCount) {
            throw new FunctionPointIndexOutOfBoundsException("Индекс " + index + " " +
                    "выходит за границы [0, " + (pointsCount - 1) + "]");
        }
        FunctionNode node;
        if (lastAccessedIndex != -1 && Math.abs(index - lastAccessedIndex) < index &&
                Math.abs(index - lastAccessedIndex) < pointsCount - index) {
            node = lastAccessedNode;
            if (index > lastAccessedIndex) {
                for (int i = lastAccessedIndex; i < index; i++) {
                    node = node.next;
                }
            } else {
                for (int i = lastAccessedIndex; i > index; i--) {
                    node = node.prev;
                }
            }
        } else {
            // Начинаем с головы
            node = head.next;
            for (int i = 0; i < index; i++) {
                node = node.next;
            }
        }

        lastAccessedNode = node;
        lastAccessedIndex = index;
        return node;
    }
```
Этот метод предназначен для оптимизации доступа к элементам списка, но его логика сложна и может быть неэффективна в некоторых случаях (например, при частом обращении к элементам в разных концах списка).

**Качество кода:**

*   **Читаемость и понятность:** Код в целом читаемый, с разумным использованием названий переменных и методов. Наличие комментариев было бы полезно для объяснения сложных алгоритмов.
*   **Простота и избегание избыточности:** Код содержит некоторую избыточность, например, дублирование проверок на корректность индекса в нескольких методах.  В `LinkedListTabulatedFunction` можно выделить общие методы для работы со списком, чтобы избежать повторения кода.
*   **Соответствие базовым рекомендациям по стилю:**  В целом, стиль кода соответствует общепринятым нормам Java.

**Общие замечания:**

*   Необходимо добавить документацию Javadoc к классам и методам.
*   Реализовать тесты для всех методов, включая граничные случаи и обработку исключений.
*   Рассмотреть возможность оптимизации производительности `LinkedListTabulatedFunction`, например, путем использования двусвязного списка с хранением ссылки на хвост списка для более быстрого добавления элементов.
*   В `ArrayTabulatedFunction` при добавлении элементов можно использовать более эффективный алгоритм изменения размера массива, например, путем увеличения размера массива в два раза.
*   В классе `ArrayTabulatedFunction` не реализована проверка на дубликаты при добавлении точки.
*   Использовать более информативные сообщения об ошибках в исключениях.

DiBoS290699 commented:
1. В setPoint записывайте копию точки, а не передаваемый объект (нарушается инкапсуляция)
2. Не используйте getNodeByIndex подряд в getFunctionValue. Сохраняйте ссылку на узел и используйте поля next и prev.
3. Добавьте в main дополнительные проверки корректной работы методов. Инициализируйте объекты TabulatedFunction  со различными значениями `y`

DiBoS290699 reviewed (commented):
# Результаты проверки текста отчета
## Рецензия на отчёт по лабораторной работе №3

**1. Краткое резюме**

Отчёт демонстрирует хорошее понимание требований задания и успешно реализует обработку исключений в классах табулированных функций, а также предоставляет реализацию на основе связного списка. Ключевым достоинством является полное выполнение поставленных задач. Главный риск/пробел – недостаточное внимание к деталям оптимизации и обоснованию выбора структуры данных (связный список vs. массив) в контексте конкретных задач.

**2. Соответствие заданию**

*   **Итог: Да**.
*   Задание 1: Изучены указанные классы исключений, что подтверждается перечислением их особенностей в отчете.
*   Задание 2: Созданы классы `FunctionPointIndexOutOfBoundsException` и `InappropriateFunctionPointException`, как указано в задании.
*   Задание 3: В классах реализована обработка исключений в конструкторах и методах, как требовалось.
*   Задание 4: Реализован класс `LinkedListTabulatedFunction` с использованием двусвязного циклического списка.
*   Задание 5: Методы `LinkedListTabulatedFunction` реализованы с учётом работы со списком и оптимизацией.
*   Задание 6: Создан интерфейс `TabulatedFunction`, оба класса реализуют его.
*   Задание 7: Проведено тестирование, включая проверку исключений.

**3. Техническая корректность (ООП и Java)**

*   Принципы ООП: Принцип инкапсуляции реализован за счет сокрытия внутренней структуры списков. Наследование используется при создании собственных исключений.
*   Иерархии/интерфейсы: Правильно определены иерархии исключений (наследование от `IndexOutOfBoundsException` и `Exception`). Интерфейс `TabulatedFunction` обеспечивает полиморфизм.
*   Исключения: Используются проверяемые исключения `IllegalArgumentException`, `FunctionPointIndexOutOfBoundsException`, `InappropriateFunctionPointException`, `IllegalStateException` в соответствующих ситуациях. Обработка исключений реализована корректно (по описанию).
*   equals-hashCode: Не применимо, так как нет пользовательских классов, требующих переопределения.
*   Потокобезопасность: Не применимо, так как нет многопоточности.

**4. Архитектура и дизайн**

*   Структура пакетов: Проект структурирован, классы находятся в пакете `functions`.
*   Паттерны: Не заявлены, что допустимо.
*   UML-диаграммы: Отсутствуют. Были бы полезны для визуализации структуры `LinkedListTabulatedFunction`.
*   Обоснование решений: Обоснование выбора двусвязного циклического списка недостаточно подробно. Не объяснено, почему именно эта структура предпочтительнее массива в контексте решаемых задач.

**5. Воспроизводимость (опционально)**

*   Не представлено — допустимо.

**6. Тестирование и результаты (опционально)**

*   Тесты: Проведено тестирование различных сценариев, включая исключения.
*   Ограничные случаи: Проверены граничные случаи (некорректные параметры конструктора, выход за границы массива).
*   Интерпретация результатов: Результаты тестов описаны в отчете.

**7. Стиль и конвенции (применимо при наличии кода)**

*   Нейминг: Имена классов и методов соответствуют Java-конвенциям.
*   Форматирование: Форматирование кода не оценивалось, так как нет полного представления.
*   Javadoc: Не оценено, так как нет кода.
*   Статический анализ: Не оценено, так как нет кода.

**8. Источники и добросовестность**

*   Ссылки: Не представлены, что допустимо.
*   Цитирование: Не применимо.
*   Заимствования: Не выявлено подозрительных заимствований.

**9. Рекомендации по улучшению**

*   Добавьте UML-диаграммы для визуализации структуры `LinkedListTabulatedFunction` и отношений между классами.
*   Более подробно обоснуйте выбор двусвязного циклического списка в качестве структуры данных для хранения точек функции. Сравните его преимущества и недостатки с массивом в контексте решаемых задач (скорость доступа, вставка/удаление элементов).
*   Добавьте комментарии к коду, объясняющие сложные алгоритмы и логику работы.
*   Рассмотрите возможность добавления модульных тестов с использованием JUnit для более полного покрытия функциональности и автоматизированной проверки.
*   Повысьте ясность изложения, избегайте повторений и формулируйте выводы более конкретно.
*   Если планируете расширять функциональность, продумайте возможность использования паттернов проектирования для повышения гибкости и расширяемости кода.


# Результаты проверки текста отчета на плагиаризм
В коде обнаружены значительные совпадения с работами seregaqt и sharllottka, в частности, фрагменты кода, демонстрирующие работу с объектами функций, вывод информации о точках и обработку исключений. Также есть сходство с кодом sr1racha777 в части обработки исключений при работе с табличными функциями.

# Результаты проверки кода
## Рецензия на лабораторную работу

### Работоспособность

В целом, код соответствует требованиям задания. Реализованы все необходимые классы исключений (`FunctionPointIndexOutOfBoundsException`, `InappropriateFunctionPointException`) и методы класса `TabulatedFunction` (как в `ArrayTabulatedFunction`, так и в `LinkedListTabulatedFunction`).

**ArrayTabulatedFunction:**

*   Конструкторы правильно выбрасывают `IllegalArgumentException` при некорректных параметрах (количество точек меньше 2, левая граница больше или равна правой).
*   Методы `getPoint()`, `setPoint()`, `getPointX()`, `setPointY()`, `deletePoint()` корректно выбрасывают `FunctionPointIndexOutOfBoundsException` при выходе за границы массива.
*   Методы `setPoint()` и `setPointX()` выбрасывают `InappropriateFunctionPointException` при попытке установить точку вне интервала соседних точек.
*   Метод `deletePoint()` выбрасывает `IllegalStateException`, если количество точек меньше 3.
*   Метод `addPoint()` выбрасывает `InappropriateFunctionPointException`, если точка с такой абсциссой уже существует.

**LinkedListTabulatedFunction:**

*   Класс `FunctionNode` реализован корректно и представляет собой элемент связного списка.
*   Все методы, аналогичные `ArrayTabulatedFunction`, корректно работают и выбрасывают соответствующие исключения.
*   Реализована оптимизация доступа к элементам списка (lastAccessedNode, lastAccessedIndex).

### Качество кода

**Общее:**

*   **Читаемость:** Код в целом читаемый, структура понятна.
*   **Названия переменных:** Названия переменных в основном понятны и отражают их назначение.
*   **Комментарии:** Комментариев достаточно, они поясняют логику работы в ключевых местах.
*   **Инкапсуляция:**  В `LinkedListTabulatedFunction` реализована хорошая инкапсуляция. Публичные методы не возвращают и не принимают ссылки на объекты `FunctionNode`.
*   **Стиль:** Код соответствует базовым рекомендациям по стилю Java (отступы, пробелы).
*   **Избыточность:** В некоторых местах есть дублирование кода между `ArrayTabulatedFunction` и `LinkedListTabulatedFunction`.  Можно было бы вынести общую логику в абстрактный класс или интерфейс (что и было сделано в задании 7).

**Проблемные участки:**

*   В `LinkedListTabulatedFunction` в методах, где нужно искать соседние точки, можно было бы использовать более понятный и компактный способ итерации по списку, не используя `getNodeByIndex()` каждый раз.
*   В `LinkedListTabulatedFunction` есть небольшое дублирование логики проверки на упорядоченность точек.

**Пример проблемного участка (LinkedListTabulatedFunction):**

```java
if (x >= getNodeByIndex(index+1).point.getX() || x <= getNodeByIndex(index-1).point.getX()) {
    throw new IllegalArgumentException("Новое значение x нарушает упорядоченность");
}
```

В этом участке несколько раз вызывается `getNodeByIndex()`, что снижает производительность.  Можно было бы сохранить ссылки на соседние узлы заранее при получении текущего узла.

**Заключение:**

Работа выполнена хорошо. Код соответствует требованиям задания, корректно работает и обладает хорошим качеством. Рекомендуется обратить внимание на замечания по оптимизации доступа к элементам списка и устранению небольшого дублирования кода.

dadppil committed:
Измененная лаба

DiBoS290699 commented:
1. В setPoint надо было всего лишь создать копию передаваемого point и записать его в нужный узел.
2. Всё ещё множество мест, где неэффективно используются возможности списка, не надо спамить в любом месте кода getNodeByIndex, вызовите его один раз в методе, если это необходимо, создайте переменную с ссылкой на этот узел и обращайтесь от него через next и prev в этом методе к соседним узлам. В showPoints например вообще можно обойтись без использования getNodeByIndex.
3. Как вы посчитали такую epsilon в виде `~2*10^(-6)`?
4. В main проверьте удаление и добавление точки в начале (с индексом 0).
5. Заменяйте старый отчёт на новый, не надо создавать второй отчёт.

dadppil committed:
Add files via upload

Добавил проверку удаления элемента под индексом 1. Значение эпсилона изначально взял из интернета, заменил на функцию которая дает это значение. Как мне добавить точку в начало списка я не понимаю, метода который добавляет по индексу у нас нет, а если пытаться добавить с меньшим значением x, то выходим за пределы интервала

dadppil commented:
Вместо 1 хотел написать 0, тут опечатка

DiBoS290699 commented:
принято
